File: tests/testthat/test-qc-diagnostics.R
fxpath <- function(name) {
  system.file("extdata", "synthetic", name, package = "MetaboSetR",
              mustWork = TRUE)
}

read_synth_prep <- function(...) {
  cd <- read_webidq(fxpath("sample_conc.xlsx"),
                    fxpath("sample_status.xlsx"))
  qc <- read_webidq_qc(fxpath("qc_conc.xlsx"),
                       fxpath("qc_status.xlsx"))
  preprocess(cd, qc, ...)
}

# ---- qc_pca_plot -----------------------------------------------------------

test_that("qc_pca_plot returns a ggplot in qc_only mode", {
  prep <- read_synth_prep()
  p <- qc_pca_plot(prep, mode = "qc_only")
  expect_s3_class(p, "ggplot")
})

test_that("qc_pca_plot returns a ggplot in with_samples mode", {
  prep <- read_synth_prep()
  p <- qc_pca_plot(prep, mode = "with_samples")
  expect_s3_class(p, "ggplot")
})

# ---- qc_permanova ----------------------------------------------------------

test_that("qc_permanova returns the adonis2 table", {
  prep <- read_synth_prep()
  res <- qc_permanova(prep)
  expect_s3_class(res, "anova")
  # adonis2 returns an anova-like data.frame; the kit_id row + Residual + Total
  expect_true(any(rownames(res) == "kit_id" |
                    rownames(res) == "Model"))
})

# ---- qc_cv_plot ------------------------------------------------------------

test_that("qc_cv_plot returns a ggplot for per_kit log", {
  prep <- read_synth_prep()
  p <- qc_cv_plot(prep)
  expect_s3_class(p, "ggplot")
})

test_that("qc_cv_plot also works in pooled CV mode", {
  prep <- read_synth_prep(cv_mode = "pooled")
  p <- qc_cv_plot(prep)
  expect_s3_class(p, "ggplot")
})

# ---- qc_lod_rate_plot ------------------------------------------------------

test_that("qc_lod_rate_plot returns a ggplot per sample/metabolite", {
  prep <- read_synth_prep()
  p1 <- qc_lod_rate_plot(prep, by = "sample")
  p2 <- qc_lod_rate_plot(prep, by = "metabolite")
  expect_s3_class(p1, "ggplot")
  expect_s3_class(p2, "ggplot")
})

test_that("qc_lod_rate_plot(by='metabolite') errors when filter was disabled", {
  prep <- read_synth_prep(apply_lod_rate_filter = FALSE)
  expect_error(qc_lod_rate_plot(prep, by = "metabolite"),
               "No `<LOD` rate data")
})

# ---- qc_filter_summary -----------------------------------------------------

test_that("qc_filter_summary tallies samples and metabolites", {
  prep <- read_synth_prep()
  s <- qc_filter_summary(prep)
  expect_s3_class(s, "data.frame")
  expect_setequal(names(s), c("item", "total", "pass", "fail"))
  expect_setequal(s$item, c("samples", "metabolites"))
  # totals should equal pass + fail per row
  expect_equal(s$total, s$pass + s$fail)
  # 6 samples, 30 metabolites in the synthetic fixture
  expect_equal(s$total[s$item == "samples"], 6L)
  expect_equal(s$total[s$item == "metabolites"], 30L)
})

File: python/tests/test_qc_diagnostics.py
"""Best-effort smoke tests for the QC diagnostics layer (CLAUDE.md §3)."""

import numpy as np
from conftest import SYNTHETIC

import metabosetr as m


def _prep():
    D = SYNTHETIC
    conc = m.read_webidq(str(D / "sample_conc.xlsx"), str(D / "sample_status.xlsx"))
    qc = m.read_webidq_qc(str(D / "qc_conc.xlsx"), str(D / "qc_status.xlsx"))
    return m.preprocess(conc, qc)


def test_filter_summary():
    prep = _prep()
    s = m.qc_filter_summary(prep)
    assert list(s["item"]) == ["samples", "metabolites"]
    # totals consistent, pass+fail == total per row
    assert (s["pass"] + s["fail"] == s["total"]).all()
    assert s.loc[s["item"] == "samples", "total"].iloc[0] == 6


def test_pca_plot_runs():
    prep = _prep()
    fig = m.qc_pca_plot(prep, mode="qc_only")
    assert fig is not None
    assert len(fig.axes) == 1


def test_permanova_runs():
    prep = _prep()
    res = m.qc_permanova(prep)  # two kits present in synthetic QC
    assert "test statistic" in res.index
    assert not np.isnan(res["p-value"])


def test_cv_and_lod_plots_run():
    prep = _prep()
    assert m.qc_cv_plot(prep) is not None
    assert m.qc_lod_rate_plot(prep, by="sample") is not None
